Adding resonance transitons for lone pairs

alongd opened the pull request:
This PR adds new resonance transitions and introduces the resonance filtration function based on heuristics.

## New resonance transitions added
### lone_pair_multiple_bond_resonance_structures
Resonance structures formed by lone electron pair - multiple bond shifts.
Example: ![image](https://user-images.githubusercontent.com/16158262/39085968-e23f28b0-4558-11e8-897d-c1b1fc871d60.png) ![image](https://user-images.githubusercontent.com/16158262/39085977-08f9c744-4559-11e8-9842-e4444e574b7b.png) ![image](https://user-images.githubusercontent.com/16158262/39085969-e6ba7070-4558-11e8-90a3-81c9c4baf653.png)

### lone_pair_radical_multiple_bond_resonance_structures
Resonance structures formed by lone electron pair - radical - multiple bond shifts.
Example: ![image](https://user-images.githubusercontent.com/16158262/39085989-385f68b8-4559-11e8-85ed-d6371b10ff4a.png) ![image](https://user-images.githubusercontent.com/16158262/39085977-08f9c744-4559-11e8-9842-e4444e574b7b.png) ![image](https://user-images.githubusercontent.com/16158262/39085991-3c39564c-4559-11e8-839e-2b7681f53c36.png)


### N5dc_resonance_structures
Resonance structures formed by radical and lone pair shifts mediated by an N5dc atom (not to be confused with `N5ddc_N5tc_resonance_structures`).
Example: ![image](https://user-images.githubusercontent.com/16158262/39086076-6c82783c-455a-11e8-9b8f-5ee336b10fb7.png) ![image](https://user-images.githubusercontent.com/16158262/39085977-08f9c744-4559-11e8-9842-e4444e574b7b.png) ![image](https://user-images.githubusercontent.com/16158262/39086076-6c82783c-455a-11e8-9b8f-5ee336b10fb7.png)
(in the above example, the two `O` atoms are "different" - this addresses issue #1293)

### Other resonance modifications
* `adjacent_resonance_structures` was renamed `ally_delocalization_resonance_structures` and elaborated for heteroatoms
* `N5dd_N5ts_resonance_structures` was renamed `N5ddc_N5tc_resonance_structures` and elaborated for sulfur
* `lone_pair_radical_resonance_structures` was elaborated for heteroatoms
* Added respective resonance and pathfinder tests

## Resonance filtration
This module contains methods for filtering a list of Molecules representing a single Species,
keeping only the representative structures. 
The rules this module follows are (by order of importance):
1. Minimum overall deviation from the Octet Rule (also elaborated for Dectet and Duodectet for hypervalance heteroatoms)
2. Additional charge separation is only important for radical species, and is kept only if it makes a new radical site in the species
3. If a structure must have charge separation, negative charges will be assigned to more electronegative atoms, and vice versa
4. Opposite charges will be as close as possible to one another, and vice versa

## Other modifications introduced by this PR
* Added the `reactive` attribute to the Molecule class. If a non-representative molecule (one that would have been filtered out if formed by `generate_resonance_structures()`) is directly formed during the `enlarge` step, the non-representative Molecule is marked an `reactive=False`, and the respresentative structures are appended to the Molecule list. If, however, this species already exists (and doesn't contain the non-representative structure), the species are merged.
* Added various tests
* Added an electronegativity dictionary to the PeriodicSystem class
* An iPython notebook for resonance generation was added

alongd commented:
This PR fails on Travis due to aug-InChI issues

alongd commented:
@mliu49, this is now ready for review.
Please advise regarding function depreciation (#1353)
It would be great if we can get this PR in the v2.2.0 release!

mliu49 commented on rmgpy/molecule/pathfinder.py:
```diff
@@ -256,67 +255,207 @@ def findAllDelocalizationPaths(atom1):
+    if (((atom.isNitrogen() or atom.isSulfur()) and atom.lonePairs in [1, 2, 3])
+                        or (atom.isOxygen() and atom.lonePairs in [2, 3])):
+        return True
+    return False
```
You could directly return the value of your boolean expression (also for gain_lone_pair).

Also, `loose` should be `lose`.

mliu49 commented on rmgpy/data/kinetics/common.py:
```diff
@@ -251,16 +255,18 @@ def independent_ids():
-            # Remake resonance structures with new labels
-            if resonance:
-                species.generate_resonance_structures(keepIsomorphic=True)
+            if all([mol.reactive for mol in species.molecule]):
```
What are potential cases where there might be an unreactive molecules when calling this method? Independent atom IDs are pretty important for degeneracy, so depending on the situation, a more comprehensive fix might be better.

alongd replied:
Molecules with a `reactive=False` flag get generated, but are not allowed to further react in families. However, we do calculate degeneracy for these reactions. My concern here was a line below that resets the molecule list (`species.molecule = [mol]`), removing all unreactive molecules. Even if we re-generate resonance structures, we won't regenerate these unreactive molecule back. I modified this that condition so that unreactive molecules are appended if needed.

mliu49 commented on rmgpy/data/kinetics/common.py:
```diff
@@ -251,16 +255,18 @@ def independent_ids():
         # IDs are already independent, generate resonance structures if needed
         for species in input_species:
-            species.generate_resonance_structures(keepIsomorphic=True)
+            if len(species.molecule) == 1:
```
This check is already done by `Species.generate_resonance_structures()`.

alongd commented on rmgpy/molecule/filtration.py:
```diff
@@ -0,0 +1,332 @@
+    # Important whenever Species.molecule[0] is expected to be used (e.g., training reactions) after generating
+    # resonance structures. However, if it was filtered out, it should be appended to the end of the list.
+    for index in xrange(len(filtered_list)):
+        if filtered_list[index].copy(deep=True).isIsomorphic(mol_list[0].copy(deep=True)):
```
Without the deep copy I get a failing test:
```
======================================================================
FAIL: Test the Molecule.calculateSymmetryNumber() on C1=C=C=1
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/alongd/Code/RMG-Py/rmgpy/molecule/symmetryTest.py", line 651, in testSymmetryNumberCyclic
    self.assertEqual(Species().fromSMILES('C1=C=C=1').getSymmetryNumber(), 6)
AssertionError: 1.0 != 6
```
This is the error we discussed in #1344. Although a simple `copy()` also sends a different object to symmetry, I still get that error unless it's a deep copy.

alongd commented on rmgpy/molecule/filtration.py:
```diff
@@ -0,0 +1,332 @@
+    """
+
+    if not any([mol.reactive for mol in filtered_list]):
+        logging.info('\n\n')
```
Yes, but then it'll print something like:
```
Error:
Error:
Error: actual error message
```
I understand it's odd, let's decide if we're keeping it or not, but we shouldn't add it to the `logging.error()`

alongd commented on rmgpy/data/kinetics/family.py:
```diff
@@ -1487,32 +1487,33 @@ def generateReactions(self, reactants, products=None, prod_resonance=True):
-            # Reverse direction (the direction in which kinetics is not defined)
-            reactionList.extend(self.__generateReactions(reactants, products=products, forward=False, prod_resonance=prod_resonance))
+
+        if all([mol.reactive for mol in reactants]):
```
Thanks, this is wrong and will be addressed in a fixup

alongd commented on ipython/generateResonanceStructures.ipynb:
```diff
@@ -0,0 +1,236 @@
+    },
+    {
+     "data": {
+      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAEMAAAAZCAIAAADlvQS1AAAABmJLR0QA/wD/AP+gvaeTAAADqUlE\nQVRYhd2YT0jbUBzHX2VuZKWSmtB2soN/wGKFzlpQpChIFaSReRp4EQZjiJdimZBDT4KTHAseXA5F\nxLIi9SBFT7WgYkVklYk6dVgQUbSS2uJoamv17ZCZxvpnmU3d2OfU3/e993u/b/L6S1oZhBD8FxT8\nxb2dTmckEpEq219zkkgkNjc3dTrd0NBQOp2WICN8dFiW7e/vx3H84OBga2vLYrFotdrp6ekc0z62\nE6/XW1ZWVlNTQ5Lk2dkZJ/p8Pp1O19LSsrGx8eDMj+dkZWWlqanJYDDMzc3t7u52dHR ...
```
I'm clearing the output cells.
I don't mind not including this notebook (tell me what you think, I'm OK with both ways). I think we previously discussed adding a resonance generation feature to the website, let's discuss it and decide whether or not to ha a filtration switch (or filter by default). It could be an excellent replacement for this notebook.

alongd commented:
Thanks for the helpful comments, @mliu49 !
I addressed them in fixup commits.
1. I'm new to autosquash, so the `xrange(len())` to `enumerate()` changes already got squashed in by mistake.
2. I think that the filtration algorithm doesn't have a significant effect on timings, I'll give some figures soon.
3. Aug-InChI was solved via 024b080170a66dc9fb2692ecfde5185d96314892. Please see the commit message. Briefly, InChI represents delocalized species, so there's no need to generate resonance structures for it.
4. Do you have particular concerns re #1329? Let's try rebasing them on a dummy branch and check.
5. I ran the tests, will add the report
6. Regarding `deflateReaction` in rmg/react,py: I implemented an error capture with `rxn.generatePairs()` in it, please let me know if and how to implement `KineticsFamily.getReactionPairs()`.

alongd commented:
Here's a comparison of number of species and resonance structure generation time for this brunch (w/ and w/o filtration) vs. master:

| Species | number of structures before filtration| time (ms) | number of structures after filtration| time inc. filtration (ms) | number of structures on master | time on master (ms) |
| ------------- | ------------- | ------------- |------------- | ------------- | ------------- | ------------- |
| NO2  | 8 | 1.03 | 2 | 1.34 | 2  | 0.23 |
| N2O  | 4 | 0.73 | 2 | 0.80  | 2 | 0.16 |
| `N(N)[N+](=O)[O-]` | 4 | 1.05 | 1 | 1.13  | 1 | 0.32 |
| `[CH2]N=O`  | 7 | 1.53 | 3 | 1.68  | 3 | 0.32 |
| `C=[C]C(=O)O[O]` | 3 | 1.52 | 2 | 1.68 | 2 | 0.25 |
| `[CH]1C=NC=N1` | 7 | 5.15 | 3 | 5.3 | 3 | 1.92 |
| `C1=C[CH]CC1([O])` | 2 | 1.62 | 2 | 1.87 | 2 | 1.0 |
| HSO | 2 | 1.13 | 2 | 1.31 | 1 | 0.03 |
| `[NH]N=S=O`  | 62 | 21.97 | 4 | 23.3 | 2 | 0.2 |

The overall time is indeed higher (about a factor of 5), but most of it is spent in generating the new structures (using new transitions not available on current master). The filtration time (`time inc. filtration` - `time` in the table) looks reasonable. For the extreme case of `[NH]N=S=O` it is

mliu49 commented on rmgpy/data/kinetics/family.py:
```diff
@@ -1464,33 +1483,65 @@ def generateReactions(self, reactants, products=None, prod_resonance=True):
-    def addReverseAttribute(self, rxn):
+        reactants = [reactant if isinstance(reactant, list) else [reactant] for reactant in reactants]
+        for reactant in reactants:
+            if not all([mol.reactive for mol in reactant]):
```
I might not have been clear about the list of lists. In the typical case, this takes reactants as `[reactantA, reactantB]`. However, it can also properly handle `[[A_res1, A_res2], [B_res1, B_res2]]`, where the sublists are lists of resonance structures for each reactant. I think I was also confused in my original comment about this part; it was correct to check that all the reactants are reactive.

Now that I think about it more though, it may be unnecessary do the check here at all. Since you already check in `__generateReactions()` where the actual generation is done, I don't think you need to change this method at all.

alongd replied:
I see. I'll revert the original change

mliu49 commented on rmgpy/species.py:
```diff
@@ -203,14 +220,28 @@ def isIdentical(self, other):
+        for species in species_list:
+            if isinstance(species, Species):
+                if self.isIsomorphic(species):
+                    return True
```
You could also do `return self.isisomorphic(species)`.

mliu49 commented on rmgpy/species.py:
```diff
@@ -203,14 +220,28 @@ def isIdentical(self, other):
+                    return True
+            else:
+                raise ValueError('Unexpected value "{0!r}" for species_list parameter;'
+                                 ' should be a List of Species objects.'.format(species))
```
Perhaps use TypeError?

alongd commented on rmgpy/molecule/molecule.py:
```diff
@@ -1033,10 +1033,7 @@ def deleteHydrogens(self):
-        for atom in self.vertices:
-            if atom.isHydrogen() and atom.label == '':
-                hydrogens.append(atom)
+        hydrogens = filter(lambda at: at.isHydrogen(), self.vertices)
```
Unintentional. I'm dropping this small commit

alongd commented:
Thanks for the comments @mliu49! I pushed additional fixups.
Currently re-running tests

mliu49 commented:
Could you convert the two assertions Codacy flagged into exceptions as well? Then I think you can go ahead and squash all of the fixups.

alongd commented:
Thanks - dealt with the two assertions, squashed and rebased

mliu49 commented on rmgpy/molecule/filtration.py:
```diff
@@ -0,0 +1,331 @@
+        for mol in filtered_list:
+            logging.info('Structure: {0}\n{1}Reactive: {2}'.format(mol.toSMILES(),mol.toAdjacencyList(),mol.reactive))
+        logging.info('\n')
+        raise AssertionError('Each species must have at least one reactive structure. Something probably went wrong'
```
I still think it would be a good idea to create a ResonanceError class to be used here.

mliu49 commented on rmgpy/molecule/inchi.py:
```diff
@@ -376,6 +376,8 @@ def _generate_minimum_resonance_isomer(mol):
 
     The metric is a sorted list with indices of the atoms that bear an unpaired electron
+
+    This function is currently depreciated
```
Spelling of `deprecated`. Also, it would be good to mention why.

I think you can add a deprecation warning similar to f7dd2f18621c8ba6a1e78f688d58b817a1b42d50 for now.

mliu49 commented on rmgpy/rmg/model.py:
```diff
@@ -262,27 +265,41 @@ def checkForExistingSpecies(self, molecule):
+                return True, True, spec
+
+        # As a last resort, check using molecule.fingerprint if the obj matches any existing species,
+        # and if it does, generate resonance structures w/o filtration and check for isomorphism
```
It doesn't seem like you're checking fingerprints as the comment says.

mliu49 commented on rmgpy/data/thermo.py:
```diff
@@ -1429,8 +1429,8 @@ def prioritizeThermo(self, species, thermoDataList):
-            indices = numpy.array([indices[i] for i in xrange(len(indices)) if species.molecule[indices[i]].reactive] +\
-                        [indices[i] for i in xrange(len(indices)) if not species.molecule[indices[i]].reactive])
+            indices = numpy.array([i for i in indices if species.molecule[indices[i]].reactive] +\
+                        [i for i in indices if not species.molecule[indices[i]].reactive])
```
I think `indices[i]` should be replaced by `i` as well.

alongd commented:
@mliu49, can I go ahead and squash?

mliu49 commented:
Yes, please go ahead. Thanks for making all the changes!

As for overall status, I don't have any more comments, the only thing left is to check RMG-tests. You mentioned that you had run it already?

alongd commented:
Here are the test results for this branch:
[resLP.txt](https://github.com/ReactionMechanismGenerator/RMG-Py/files/2065336/resLP.txt)

There were no changes for eg1, eg3, eg6, eg7, and solvent_hexane

------------------------
In eg5, order in which networks were generated was different, so we get things like:
```
Non-identical kinetics!
original:
rxn: C=C(16) + C=CC(30) => [CH2]CC([CH2])C(66)		origin: PDepNetwork #40
tested:
rxn: C=CC(30) + C=C(16) => [CH2]CC([CH2])C(66)		origin: PDepNetwork #40
k(1bar)|300K   |400K   |500K   |600K   |800K   |1000K  |1500K  |2000K  

k(T):  | -29.36| -21.27| -16.40| -13.13|  -8.99|  -6.47|  -3.07|  -1.35
k(T):  | -58.17| -42.17| -32.52| -26.04| -17.83| -12.83|  -6.08|  -2.69

Kinetics: Chebyshev(coeffs=[[-9.141,2.483,-0.01289,-0.007827],[14.85,0.01142,0.008461,0.005105],[0.1403,0.0008223,0.0006267,0.0003947],[0.03315,0.0005011,0.0003743,0.0002288],[0.001036,0.0003139,0.0002346,0.0001435],[-0.007855,0.0002016,0.0001507,9.213e-05]], kunits='cm^3/(mol*s)', Tmin=(300,'K'), Tmax=(3000,'K'), Pmin=(0.001,'atm'), Pmax=(98.692,'atm'))
Kinetics: Chebyshev(coeffs=[[-23.02,-0.0164,-0.0122,-0.007413],[29.41,0.01176,0.008717,0.005266],[0.2732,0.001358,0.001026,0.0006379],[0.05007,0.0007712,0.0005766,0.0003528],[-0.007379,0.0003885,0.0002909,0.0001783],[-0.02062,0.0001704,0.0001278,7.853e-05]], kunits='cm^3/(mol*s)', Tmin=(300,'K'), Tmax=(3000,'K'), Pmin=(0.001,'atm'), Pmax=(98.692,'atm'))
Identical kinetics comments

Non-identical kinetics!
original:
rxn: C=C(16) + C=CC(30) => [CH2]CC([CH2])C(66)		origin: PDepNetwork #39
tested:
rxn: C=CC(30) + C=C(16) => [CH2]CC([CH2])C(66)		origin: PDepNetwork #39
k(1bar)|300K   |400K   |500K   |600K   |800K   |1000K  |1500K  |2000K  

k(T):  | -58.17| -42.17| -32.52| -26.04| -17.83| -12.83|  -6.08|  -2.69
k(T):  | -29.36| -21.27| -16.40| -13.13|  -8.99|  -6.47|  -3.07|  -1.35

Kinetics: Chebyshev(coeffs=[[-23.02,-0.0164,-0.0122,-0.007413],[29.41,0.01176,0.008717,0.005266],[0.2732,0.001358,0.001026,0.0006379],[0.05007,0.0007712,0.0005766,0.0003528],[-0.007379,0.0003885,0.0002909,0.0001783],[-0.02062,0.0001704,0.0001278,7.853e-05]], kunits='cm^3/(mol*s)', Tmin=(300,'K'), Tmax=(3000,'K'), Pmin=(0.001,'atm'), Pmax=(98.692,'atm'))
Kinetics: Chebyshev(coeffs=[[-9.141,2.483,-0.01289,-0.007827],[14.85,0.01142,0.008461,0.005105],[0.1403,0.0008223,0.0006268,0.0003947],[0.03315,0.0005011,0.0003743,0.0002288],[0.001035,0.0003139,0.0002346,0.0001435],[-0.007855,0.0002016,0.0001507,9.213e-05]], kunits='cm^3/(mol*s)', Tmin=(300,'K'), Tmax=(3000,'K'), Pmin=(0.001,'atm'), Pmax=(98.692,'atm'))
Identical kinetics comments
```
@mliu49, I was wrong yesterday to think that only the first network reaction instance gets included in the model, RMG seems to be doing the right thing here.

------------------------
In the NC test, this branch corrected the representative structure of NNOH, so the test treats them as a different species.

------------------------
The methane test (with reactive N2) had some discrepancies, with 4 species not included in the new model, and 2 additional species. Some kinetics are different due to two reasons:
- different sorting of resonance structures, e.g.:
```
Non-identical kinetics!
original:
rxn: CH2CHO(49) + oxygen(3) <=> HO2(8) + CH2CO(25)		origin: Disproportionation
tested:
rxn: CH2CHO(49) + oxygen(3) <=> CH2CO(25) + HO2(8)		origin: Disproportionation
k(1bar)|300K   |400K   |500K   |600K   |800K   |1000K  |1500K  |2000K  

k(T):  |   1.48|   2.55|   3.20|   3.63|   4.17|   4.49|   4.92|   5.13
k(T):  |  -3.79|  -1.32|   0.16|   1.15|   2.38|   3.12|   4.11|   4.60

Kinetics: Arrhenius(A=(6.02974e+11,'cm^3/(mol*s)'), n=0, Ea=(5.908,'kcal/mol'), T0=(1,'K'), comment="""Estimated using template [O2b;XH_s_Rrad] for rate rule [O2b;COpri_Csrad]
Kinetics: Arrhenius(A=(1.2044e+12,'cm^3/(mol*s)'), n=0, Ea=(13.55,'kcal/mol'), T0=(1,'K'), comment="""Estimated using template [O2b;Cdpri_Rrad] for rate rule [O2b;Cdpri_Orad]
kinetics: Estimated using template [O2b;XH_s_Rrad] for rate rule [O2b;COpri_Csrad]
kinetics: Estimated using template [O2b;Cdpri_Rrad] for rate rule [O2b;Cdpri_Orad]
```
- Different reaction enthalpy change when a more representative resonance structure was found with lower H298, e.g:
```
Non-identical kinetics!
original:
rxn: O(T)(4) + N2O(34) <=> [O-][N+]=[N+][O-](105)		origin: R_Addition_MultipleBond
tested:
rxn: N2O(34) + O(T)(4) <=> [O][N]N=O(105)		origin: R_Addition_MultipleBond
k(1bar)|300K   |400K   |500K   |600K   |800K   |1000K  |1500K  |2000K  

k(T):  |  -6.42|  -4.08|  -2.80|  -2.03|  -1.22|  -0.86|  -0.64|  -0.75
k(T):  | -42.05| -30.80| -24.18| -19.85| -14.58| -11.55|  -7.77|  -6.09

Kinetics: Arrhenius(A=(9.99e+25,'cm^3/(mol*s)'), n=-5.73, Ea=(16.784,'kcal/mol'), T0=(1,'K'), comment="""Estimated using an average for rate rule [N3t_R;O_atom_triplet]
Kinetics: Arrhenius(A=(9.99e+25,'cm^3/(mol*s)'), n=-5.73, Ea=(65.694,'kcal/mol'), T0=(1,'K'), comment="""Estimated using an average for rate rule [N3t_R;O_atom_triplet]
kinetics: Estimated using an average for rate rule [N3t_R;O_atom_triplet]
kinetics: Estimated using an average for rate rule [N3t_R;O_atom_triplet]
```
The updated thermo, however, isn't alwas better. e.g.:
```
Non-identical thermo!
original:	[O-][N+]=[N+][O-]
tested:	[O][N]N=O
Hf(300K)  |S(300K)   |Cp(300K)  |Cp(400K)  |Cp(500K)  |Cp(600K)  |Cp(800K)  |Cp(1000K) |Cp(1500K) 
     57.60|     78.45|     16.64|     17.44|     18.01|     18.75|     21.18|     23.73|     25.71
    104.32|     27.82|      4.01|      3.87|      4.01|      4.23|      4.46|      4.37|      4.55
thermo: Thermo group additivity estimation: group(O2s-CsCs) + group(O2s-CsCs) + group(N3s-CsHH) + group(N3s-CsHH) + radical(NHJ_C) + radical(NHJ_C)
thermo: Thermo group additivity estimation: group(N3s-CsHH) + radical(CsOJ) + radical(NHJ_C)
```
[O-][N+]=[N+][O-] turned into a more representative form, [O][N]N=O, but now only one N group is recognized. It's a step in the right direction, but GAV must be update.
There were many changes to the edge.

----------------
The Methylcyclohexane (MCH) test had some difference in species, probably die to the only core reaction kinetics difference in the core:
```
Non-identical kinetics!
original:
rxn: [CH2]C=CC=C[CH2](406) <=> C=CC1C=CC1(530)		origin: Birad_recombination
tested:
rxn: [CH2]C=CC=C[CH2](406) <=> C=CC1C=CC1(530)		origin: Birad_recombination
k(1bar)|300K   |400K   |500K   |600K   |800K   |1000K  |1500K  |2000K  

k(T):  |  11.29|  11.62|  11.82|  11.95|  12.11|  12.21|  12.34|  12.41
k(T):  |  10.31|  10.64|  10.82|  10.94|  11.08|  11.16|  11.25|  11.29

Kinetics: Arrhenius(A=(4e+12,'s^-1'), n=0, Ea=(1.8,'kcal/mol'), T0=(1,'K'), comment="""Estimated using template [Rn;C_rad_out_H/OneDe;Cpri_rad_out_2H] for rate rule [R4_SDS;C_rad_out_H/OneDe;Cpri_rad_out_2H]
Kinetics: Arrhenius(A=(3.24e+12,'s^-1'), n=-0.305, Ea=(1.98,'kcal/mol'), T0=(1,'K'), comment="""Estimated using template [R4;C_rad_out_2H;Cpri_rad_out_single] for rate rule [R4_SDS;C_rad_out_2H;Cpri_rad_out_H/OneDe]
kinetics: Estimated using template [Rn;C_rad_out_H/OneDe;Cpri_rad_out_2H] for rate rule [R4_SDS;C_rad_out_H/OneDe;Cpri_rad_out_2H]
kinetics: Estimated using template [R4;C_rad_out_2H;Cpri_rad_out_single] for rate rule [R4_SDS;C_rad_out_2H;Cpri_rad_out_H/OneDe]
```
Seems like label *1 and *2 switched her., I'm not that familiar with this family, but probably due to a different sorting of the resonance structures.

---------------
The new H2S test crashed because of incorrect parsing of library names, I'll try to deal with it separately.

mliu49 commented:
Thanks for looking at the tests! I think this is ready to be merged. The final thing (which I almost forgot about) is whether or not to keep `generateResonanceStructures.ipynb`. The main reason that I'm hesitant is that I don't think it's necessarily useful for users. How would you feel about relocating it to the Work-In-Progress_scripts repo?

alongd commented:
Notebook removed. It should be ready to go now.
